src/powerquery-language-services/analysis.ts

```diff
@@ -153,24 +137,72 @@ abstract class AnalysisBase implements Analysis {
             return LanguageServiceUtils.EmptySignatureHelp;
         }
 
-        // TODO: add tracing/logging to the catch()
-        const librarySignatureHelp: Promise<SignatureHelp | null> = this.librarySymbolProvider
-            .getSignatureHelp(context)
-            .catch(() => {
-                // tslint:disable-next-line: no-null-keyword
-                return null;
-            });
-
-        const [libraryResponse] = await Promise.all([librarySignatureHelp]);
-
-        return libraryResponse ?? LanguageServiceUtils.EmptySignatureHelp;
+        // Result priority is based on the order of the symbol providers
+        return AnalysisBase.resolveProviders(
+            AnalysisBase.createSignatureHelpCalls(context, [
+                this.localSymbolProvider,
+                this.environmentSymbolProvider,
+                this.librarySymbolProvider,
+            ]),
+            LanguageServiceUtils.EmptySignatureHelp,
+        );
     }
 
     public abstract dispose(): void;
 
     protected abstract getLexerState(): WorkspaceCache.LexerCacheItem;
     protected abstract getText(range?: Range): string;
 
+    private static createCompletionItemCalls(
+        context: CompletionItemProviderContext,
+        providers: CompletionItemProvider[],
+    ): Promise<CompletionItem[]>[] {
+        // TODO: add tracing to the catch case
+        return providers.map(provider =>
+            provider.getCompletionItems(context).catch(() => {
+                return LanguageServiceUtils.EmptyCompletionItems;
+            }),
+        );
+    }
+
+    private static createHoverCalls(
+        context: HoverProviderContext,
+        providers: HoverProvider[],
+    ): Promise<Hover | null>[] {
+        // TODO: add tracing to the catch case
+        return providers.map(provider =>
+            provider.getHover(context).catch(() => {
+                // tslint:disable-next-line: no-null-keyword
+                return null;
```

Comment:
I'm assuming all of these nulls are due to types coming from language server types. Should we disable tslint no-null-keyword for this file or have several disable-next-line?

Reply:
Yeah, the nulls were coming from the language service types, although looking closer I think we might be able to get away with using undefined instead. 

I'm putting this off since we need to move from tslint to eslint and update a number of our rules.